### ISSUE TITLE
[a11y] Toast messages not announced by screen reader and not dismissible by keyboard controls

### ISSUE THREAD

**Issue by melanie-exygy** [civiform/civiform#7749]
**Describe the bug**
Toast messages in CiviForm are not announced by the screen reader so vision-impaired users will not know when an alert is shown.  They are also not dismissible via keyboard controls.

**To Reproduce**

1. Log in as a CiviForm Admin.
2. Add an eligibility criteria to a program.
3. Publish your changes.
4. Logout
5. From the homepage, apply to the program you edited.
6. Answer the question with the eligibility criteria in a way that doesn't meet eligibility.
7. Click "Save and Next"
8. Turn on the screen reader (Command + F5 on a Mac)
9. Click "Go back and edit".
10. Note that the screen reader doesn't read the toast message that says the user in ineligible.
11. Also note that the toast exit button isn't accessible through keyboard controls.

**Expected behavior**
The screen reader should announce the content of the toast message and the toast dismiss button should be accessible via keyboard controls.

**Desktop (please complete the following information):**

- OS: iOS
- Browser: Chrome
- Screen reader: VoiceOver

**Additional context**
The toast messages are contained in an `aria-live` region and have the correct roles (i.e. `role="alert"`).  Based on the note [here](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions), I suspect that the aria-live regions aren't being recognized by the screen reader because they are not changing dynamically (through JavaScript).  Rather, the toasts are rendered on page load and they don't change.

Here is guidance from the UX design lead at Exygy:

1.  Toasts are intended to be small interruptions to your visitors or users, so to help those with screen readers and similar assistive technologies, you should wrap your toasts in an aria-live[ region](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions).
2.  You also need to adapt the role and aria-live level depending on the content. If it’s an important message like an error, use role="alert" aria-live="assertive", otherwise use role="status" aria-live="polite" attributes.
3.  Ensure that the toast can be dismissed using keyboard commands.
[Status messages](https://www.w3.org/WAI/WCAG21/Understanding/status-messages.html)
[Aria techniques](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA19)
[Aria live regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions (edited))

**Comment by melanie-exygy** [comment]
@myfriendgu 

**List of remaining applicant-side toasts**
1. "Error saving application": Gray info toast. Doesn't auto-dismiss. Dismissible via button. See screenshot above. Triggered from the application review page when there is an `ApplicationSubmissionException` during submit. There's no natural way to provoke this from a healthy environment.
2. "Error: There's been an update to the application, please review and answer questions to proceed": Red error toast. Doesn't auto-dismiss. Not dismissible. Triggered from the application review page when there is an `ApplicationOutOfDateException` during submit. According to the code comment this happens when: ""An application may be submitted but incomplete if the application view with submit button contains stale data that has changed visibility conditions."
3. "Session successfully extended": Green success toast. Auto-dismisses after 3 seconds.  Dismissible via button. Triggered when you click "Extend Session" in the inactivity warning modal.  Triggered when you click "Extend Session" in the inactivity warning modal.
4. "Failed to extend session": Red error toast. Auto-dismisses after 3 seconds.  Dismissible via button. Triggered when you click "Extend Session" in the inactivity warning modal and there's a network error in extending the session.  Triggered when you click "Extend Session" in the inactivity warning modal and there's a network error in extending the session.
5. "Your session has ended.": Green success toast. Auto-dismisses after 5 seconds. Dismissible via button. Click "end your session".

**Comment by melanie-exygy** [comment]
Our plan is to replace these 5 toasts with alerts.

**Comment by rockycodes** [comment]
Charlotte flagged the green toast as an accessibility concern for contrast. If we are getting rid of the toasts in favor of alerts, that will resolve the issue. We'll just need to make sure the contrast meets these guidelines. 

<img width="611" height="307" alt="Image" src="https://github.com/user-attachments/assets/47898c3a-899a-41e4-be63-6d37a89eb1ab" />

Remediation: According to WCAG 2.1 Level AA:
• Text less than 18 point or bold text less than 14 point must have a luminosity
contrast ratio of 4.5:1 or more.
• Text 18 point or larger or bold text 14 point or larger must have a luminosity
contrast ratio of 3:1 or more.

https://www.w3.org/TR/WCAG21/#contrast-minimum

**Comment by myfriendgu** [comment]
Hi @melanie-exygy and @rockycodes, I agree that these toast messages need to get converted to [alerts](https://www.figma.com/design/aHJ6bH2vR3CNK2DcJHAeOj/-Admin-DRAFT--CiviForm--Components--%E2%80%A8Styles--Templates?node-id=956-5455). I've marked the different type that they should be redone as in each. I've also added some additional questions/thoughts I have regarding the copy for some alerts:

> 1. "Error saving application": Gray info toast. Doesn't auto-dismiss. Dismissible via button. See screenshot above. Triggered from the application review page when there is an ApplicationSubmissionException during submit. There's no natural way to provoke this from a healthy environment.

^ This needs to get redone as an error alert (red). Ideally any alert we have gives vital information and has a step to resolve if it's an error. This one seems like an error but doesn't give any resolution... is there anything the user can do for this to remedy?

If not and the only thing the user can do is close out, then I recommend going with a warning alert (yellow) with the x, but I don't love that.


> 2. "Error: There's been an update to the application, please review and answer questions to proceed": Red error toast. Doesn't auto-dismiss. Not dismissible. Triggered from the application review page when there is an ApplicationOutOfDateException during submit. According to the code comment this happens when: ""An application may be submitted but incomplete if the application view with submit button contains stale data that has changed visibility conditions."

^ This one can get redone as an error alert (red). This one seems straight forward; however, do you think we need to have users refresh the page _then_ review and answer questions to proceed?


> 3. "Session successfully extended": Green success toast. Auto-dismisses after 3 seconds. Dismissible via button. Triggered when you click "Extend Session" in the inactivity warning modal. Triggered when you click "Extend Session" in the inactivity warning modal.

^ This one can get redone as a success alert (green).


> 4. "Failed to extend session": Red error toast. Auto-dismisses after 3 seconds. Dismissible via button. Triggered when you click "Extend Session" in the inactivity warning modal and there's a network error in extending the session. Triggered when you click "Extend Session" in the inactivity warning modal and there's a network error in extending the session.

^ This one can get redone as an error alert (red). Similar to the "Error saving application" alert, ideally we have a next step for users (unless that's a security risk?). Is there any more information we can give?


> 5. "Your session has ended.": Green success toast. Auto-dismisses after 5 seconds. Dismissible via button. Click "end your session".

^ This one can get redone as a success alert (green).

**Comment by melanie-exygy** [comment]
Hey @myfriendgu , continuing the discussion on converting toasts to alerts.  One general question I have is where on the page these alerts should be displayed.  Would you mind adding instructions for that?  @rockycodes , I'd especially love your perspective on #1 and #4.

> Hi [@melanie-exygy](https://github.com/melanie-exygy) and [@rockycodes](https://github.com/rockycodes), I agree that these toast messages need to get converted to [alerts](https://www.figma.com/design/aHJ6bH2vR3CNK2DcJHAeOj/-Admin-DRAFT--CiviForm--Components--%E2%80%A8Styles--Templates?node-id=956-5455). I've marked the different type that they should be redone as in each. I've also added some additional questions/thoughts I have regarding the copy for some alerts:
> 
> > 1. "Error saving application": Gray info toast. Doesn't auto-dismiss. Dismissible via button. See screenshot above. Triggered from the application review page when there is an ApplicationSubmissionException during submit. There's no natural way to provoke this from a healthy environment.
> 
> ^ This needs to get redone as an error alert (red). Ideally any alert we have gives vital information and has a step to resolve if it's an error. This one seems like an error but doesn't give any resolution... is there anything the user can do for this to remedy?
> 
> If not and the only thing the user can do is close out, then I recommend going with a warning alert (yellow) with the x, but I don't love that.
> 

^ Since there isn't anything the user can do to remedy this, I suggest that we handle this like any other internal server error, by sending them to the error page, which looks like this:  

<img width="690" height="189" alt="Image" src="https://github.com/user-attachments/assets/f3db54ba-0892-4b15-b0d0-4f94ea207471" />
  

> > 2. "Error: There's been an update to the application, please review and answer questions to proceed": Red error toast. Doesn't auto-dismiss. Not dismissible. Triggered from the application review page when there is an ApplicationOutOfDateException during submit. According to the code comment this happens when: ""An application may be submitted but incomplete if the application view with submit button contains stale data that has changed visibility conditions."
> 
> ^ This one can get redone as an error alert (red). This one seems straight forward; however, do you think we need to have users refresh the page _then_ review and answer questions to proceed?
> 
^ That's a good consideration, but the user shouldn't need to refresh since the page just reloaded to show them the error message.

> > 3. "Session successfully extended": Green success toast. Auto-dismisses after 3 seconds. Dismissible via button. Triggered when you click "Extend Session" in the inactivity warning modal. Triggered when you click "Extend Session" in the inactivity warning modal.
> 
> ^ This one can get redone as a success alert (green).
> 
> > 4. "Failed to extend session": Red error toast. Auto-dismisses after 3 seconds. Dismissible via button. Triggered when you click "Extend Session" in the inactivity warning modal and there's a network error in extending the session. Triggered when you click "Extend Session" in the inactivity warning modal and there's a network error in extending the session.
> 
> ^ This one can get redone as an error alert (red). Similar to the "Error saving application" alert, ideally we have a next step for users (unless that's a security risk?). Is there any more information we can give?
> 
^ I don't think we're set up to restore a guest user's session after it's closed, so there isn't much we can offer as next steps.  For a registered user, we could probably say something like "Please log in again to continue your application."

> > 5. "Your session has ended.": Green success toast. Auto-dismisses after 5 seconds. Dismissible via button. Click "end your session".
> 
> ^ This one can get redone as a success alert (green).

**Comment by myfriendgu** [comment]
1. Error alert that notifies users that an error has occurred, that their app has NOT been submitted, to wait a couple minutes to submit again, if they keep getting the error to email [support] with the error #

2. Treat as uswds error alert: "Error: There's been an update to the application, please review and answer questions to proceed"

3. Treat as uswds success alert: "Session successfully extended"

4. Melanie will replicate this to confirm which page this toast message displays / which page the user ultimately ends up on. Melanie also to confirm if a user were to log in, whether they're sent back to the same place they left off at AND if the info they filled out remains.

5. Treat as uswds success alert: "Your session has ended."

**Comment by melanie-exygy** [comment]
Test results for #4:

- If they’re a guest and log in after extending the session fails:
    - They are kept on the same page after logging in
    - Their data is retained
    - The act of logging in extends their session because they are no longer inactive
- If they're a guest and create a new account after extending the session fails:
    - Same behavior as above

**Comment by myfriendgu** [comment]
^ separate tickets to convert for all five toast messages to alerts have been created and cross-linked
